File: core/src/rustyice/editor/WidgetBuilders/PropertyWidgetBuilder.java
package rustyice.editor.widgetbuilders;

import com.badlogic.gdx.scenes.scene2d.ui.WidgetGroup;

import java.lang.reflect.Method;

/**
 * @author gabek
 */
public abstract class PropertyWidgetBuilder {
    private Object component;
    private Method getter;
    private Method setter;

    private String title;

    public PropertyWidgetBuilder(Object component, String title){
        this.component = component;
        this.title = title;
    }

    public void setSetter(Method setter){
        this.setter = setter;
    }

    public void setGetter(Method getter){
        this.getter = getter;
    }

    public Object getComponent() {
        return component;
    }

    public void setComponent(Object component) {
        this.component = component;
    }

    public Method getGetter() {
        return getter;
    }

    public Method getSetter() {
        return setter;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public abstract void addMethod(Method method);

    public abstract void buildWidgets(WidgetGroup group);
}

File: core/src/rustyice/editor/widgetbuilders/PropertyWidgetBuilder.kt
package rustyice.editor.widgetbuilders;

import com.badlogic.gdx.scenes.scene2d.ui.WidgetGroup;
import com.kotcrab.vis.ui.widget.VisTable;

import java.lang.reflect.Method;
import kotlin.reflect.KMutableProperty

/**
 * @author gabek
 */
abstract class PropertyWidgetBuilder() {
    lateinit var title: String private set
    lateinit var component: Any private set
    lateinit var property: KMutableProperty<*> private set

    fun init(title: String, component: Any, property: KMutableProperty<*>): PropertyWidgetBuilder{
        this.title = title
        this.component = component
        this.property = property
        return this
    }

    abstract fun buildWidgets(): VisTable
}
